packages/mui-joy/src/Badge/Badge.tsx

```diff
@@ -0,0 +1,387 @@
+    ],
+  };
+
+  return composeClasses(slots, getBadgeUtilityClass, {});
+};
+
+const BadgeRoot = styled('span', {
+  name: 'MuiBadge',
+  slot: 'Root',
+  overridesResolver: (props, styles) => styles.root,
+})<{ ownerState: BadgeProps }>(() => ({
+  position: 'relative',
+  display: 'inline-flex',
+  // For correct alignment with the text.
+  verticalAlign: 'middle',
+  flexShrink: 0,
+}));
+
+const BadgeBadge = styled('span', {
+  name: 'MuiBadge',
+  slot: 'Badge',
+  overridesResolver: (props, styles) => styles.badge,
+})<{ ownerState: BadgeProps }>(({ theme, ownerState }) => {
+  return [
+    {
+      ...(ownerState.size === 'sm' && {
+        '--Badge-minWidth': '1rem',
+        '--Badge-height': '1rem',
+        '--Badge-borderRadius': '0.5rem',
+        '--Badge-padding': '0.333rem',
+        '--Badge-fontSize': theme.vars.fontSize.sm,
+      }),
+      ...(ownerState.size === 'md' && {
+        '--Badge-minWidth': '1.25rem',
+        '--Badge-height': '1.25rem',
+        '--Badge-borderRadius': '0.625rem',
+        '--Badge-padding': '0.416rem',
+        '--Badge-fontSize': theme.vars.fontSize.md,
+      }),
+      ...(ownerState.size === 'lg' && {
+        '--Badge-minWidth': '1.5rem',
+        '--Badge-height': '1.5rem',
+        '--Badge-borderRadius': '0.75rem',
+        '--Badge-padding': '0.5rem',
+        '--Badge-fontSize': theme.vars.fontSize.lg,
+      }),
+      display: 'flex',
+      flexDirection: 'row',
+      flexWrap: 'wrap',
+      justifyContent: 'center',
+      alignContent: 'center',
+      alignItems: 'center',
+      position: 'absolute',
+      boxSizing: 'border-box',
+      fontFamily: theme.vars.fontFamily.body,
+      fontWeight: theme.vars.fontWeight.sm,
+      lineHeight: 1,
+      padding: '0 var(--Badge-padding)',
+      fontSize: 'var(--Badge-fontSize)',
+      minWidth: 'var(--Badge-minWidth)',
+      height: 'var(--Badge-height)',
+      borderRadius: 'var(--Badge-borderRadius)',
+      zIndex: 1, // Render the badge on top of potential ripples.
+      transition: 'transform 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms',
+      ...(ownerState.anchorOrigin!.vertical === 'top' &&
+        ownerState.anchorOrigin!.horizontal === 'right' &&
+        ownerState.location === 'outside' && {
+          top: 0,
+          right: 0,
+          transform: 'scale(1) translate(50%, -50%)',
+          transformOrigin: '100% 0%',
+          [`&.${badgeClasses.invisible}`]: {
+            transform: 'scale(0) translate(50%, -50%)',
+          },
+        }),
+      ...(ownerState.anchorOrigin!.vertical === 'bottom' &&
+        ownerState.anchorOrigin!.horizontal === 'right' &&
+        ownerState.location === 'outside' && {
+          bottom: 0,
+          right: 0,
+          transform: 'scale(1) translate(50%, 50%)',
+          transformOrigin: '100% 100%',
+          [`&.${badgeClasses.invisible}`]: {
+            transform: 'scale(0) translate(50%, 50%)',
+          },
+        }),
```

Comment:
I guess there should be a better way to simplify this logic. Let me try it first and if I could not find it, we can move forward with this.

Reply:
Found it.